src/Vault.sol

```diff
@@ -0,0 +1,398 @@
+    function _deposit(address asset_, address caller, address receiver, uint256 assets, uint256 shares) internal {
+        _getVaultStorage().totalAssets += assets;
+
+        SafeERC20.safeTransferFrom(IERC20(asset_), caller, address(this), assets);
+
+        _mint(receiver, shares);
+
+        emit Deposit(caller, receiver, assets, shares);
+    }
+
+    // QUESTION: How might we
+    function _withdraw(address caller, address receiver, address owner, uint256 assets_, uint256 shares) internal {
+        // reduce the totalAssets QUESTION: Should this go in the deposit function:
+        _getVaultStorage().totalAssets -= assets_;
+
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+
+        // Check if the contract has enough balance
+        uint256 contractBalance = IERC20(asset()).balanceOf(address(this));
+        if (contractBalance < assets_) {
+            uint256 shortfall = assets_ - contractBalance;
+
+            // pull base asset shortfall from the buffer strategy
+            IStrategy(_getVaultStorage().bufferStrategy).withdraw(shortfall, address(this), address(this));
+        }
+
+        _burn(owner, shares);
+        SafeERC20.safeTransfer(IERC20(asset()), receiver, assets_);
+
+        emit Withdraw(caller, receiver, owner, assets_, shares);
+    }
+
+    function _decimalsOffset() internal pure returns (uint8) {
+        return 0;
+    }
+
+    function _getVaultStorage() internal pure returns (VaultStorage storage $) {
+        assembly {
+            $.slot := 0x22cdba5640455d74cb7564fb236bbbbaf66b93a0cc1bd221f1ee2a6b2d0a2427
+        }
+    }
+
+    function _getAssetStorage() internal pure returns (AssetStorage storage $) {
+        assembly {
+            $.slot := 0x2dd192a2474c87efcf5ffda906a4b4f8a678b0e41f9245666251cfed8041e680
+        }
+    }
+
+    function _getStrategyStorage() internal pure returns (StrategyStorage storage $) {
+        assembly {
+            $.slot := 0x36e313fea70c5f83d23dd12fc41865566e392cbac4c21baf7972d39f7af1774d
+        }
+    }
+
+    //// ADMIN ////
+
+    function setRateProvider(address rateProvider_) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (rateProvider_ == address(0)) revert ZeroAddress();
+        _getVaultStorage().rateProvider = rateProvider_;
+        emit SetRateProvider(rateProvider_);
+    }
+
+    function setBufferStrategy(address bufferStrategy_) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (bufferStrategy_ == address(0)) revert ZeroAddress();
+        StrategyStorage storage strategyStorage = _getStrategyStorage();
+        if (!strategyStorage.strategies[bufferStrategy_].active) revert InvalidStrategy();
+        _getVaultStorage().bufferStrategy = bufferStrategy_;
+        emit SetBufferStrategy(bufferStrategy_);
+    }
+
+    function addAsset(address asset_, uint8 decimals_) external onlyRole(DEFAULT_ADMIN_ROLE) {
```

Comment:
we can consider reading asset.decimals() directly from asset, grand majority support that function

Reply:
my concern is with handling USDC, which I suspect we will with ynUSDx

Reply:
valid concern

https://etherscan.io/address/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48#readProxyContract

USDC has 6 decimals. so does USDT and Paypal USD etc.

We'll have to decide what ynUSDx has - i'd go for 18 like DAI, Ethena and the Defi Stables. It would still work to have our vaults have 18 decimals and assets with different decimals so it

I noticed one gotcha here though: 

https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/723f8cab09cdae1aca9ec9cc1cfa040c2d4b06c1/contracts/token/ERC20/extensions/ERC4626Upgradeable.sol#L98C30-L98C43

OZ default ERC4626 implementation takes the decimal count from whatever you configure as the asset - we'd need to override that behaviour

Reply:
Yes, this will need to get added to the scenario testing for a USD vault.